File: systemstubs/src/main/java/android/app/smartspace/uitemplatedata/Icon.java
package android.app.smartspace.uitemplatedata;

import android.os.Parcel;
import android.os.Parcelable;

import androidx.annotation.Nullable;

public class Icon implements Parcelable {

    public static final class Builder {
        public Builder(android.graphics.drawable.Icon icon) {
            throw new RuntimeException("Stub!");
        }

        public Icon build() {
            throw new RuntimeException("Stub!");
        }

        public Builder setContentDescription(@Nullable CharSequence contentDescription) {
            throw new RuntimeException("Stub!");
        }

        public Builder setShouldTint(boolean shouldTint) {
            throw new RuntimeException("Stub!");
        }
    }

    public CharSequence getContentDescription() {
        throw new RuntimeException("Stub!");
    }

    public android.graphics.drawable.Icon getIcon() {
        throw new RuntimeException("Stub!");
    }

    public boolean shouldTint() {
        throw new RuntimeException("Stub!");
    }

    protected Icon(Parcel in) {
        throw new RuntimeException("Stub!");
    }

    public static final Creator<Icon> CREATOR = new Creator<Icon>() {
        @Override
        public Icon createFromParcel(Parcel in) {
            throw new RuntimeException("Stub!");
        }

        @Override
        public Icon[] newArray(int size) {
            throw new RuntimeException("Stub!");
        }
    };

    @Override
    public int describeContents() {
        throw new RuntimeException("Stub!");
    }

    @Override
    public void writeToParcel(Parcel parcel, int i) {
        throw new RuntimeException("Stub!");
    }
}

File: sdk-core/src/main/java/com/kieronquinn/app/smartspacer/sdk/model/uitemplatedata/Icon.kt
package com.kieronquinn.app.smartspacer.sdk.model.uitemplatedata

import android.os.Bundle
import android.os.Parcel
import android.os.Parcelable
import androidx.annotation.RestrictTo
import androidx.core.os.bundleOf
import com.kieronquinn.app.smartspacer.sdk.utils.getParcelableCompat
import kotlinx.parcelize.Parcelize
import android.graphics.drawable.Icon as AndroidIcon

@Parcelize
data class Icon(
    val icon: AndroidIcon,
    var contentDescription: CharSequence? = null,
    val shouldTint: Boolean = true
): Parcelable {

    companion object {
        private const val KEY_ICON = "icon"
        private const val KEY_CONTENT_DESCRIPTION = "content_description"
        private const val KEY_SHOULD_TINT = "should_tint"
    }

    @RestrictTo(RestrictTo.Scope.LIBRARY)
    constructor(bundle: Bundle): this(
        bundle.getParcelableCompat<AndroidIcon>(KEY_ICON, AndroidIcon::class.java)!!,
        bundle.getCharSequence(KEY_CONTENT_DESCRIPTION),
        bundle.getBoolean(KEY_SHOULD_TINT)
    )

    fun toBundle(): Bundle {
        return bundleOf(
            KEY_ICON to icon,
            KEY_CONTENT_DESCRIPTION to contentDescription,
            KEY_SHOULD_TINT to shouldTint
        )
    }

    fun clone(): Icon {
        return copy(icon = icon.copy())
    }

    private fun AndroidIcon.copy(): AndroidIcon {
        val parcel = Parcel.obtain()
        writeToParcel(parcel, 0)
        parcel.setDataPosition(0)
        val icon = AndroidIcon.CREATOR.createFromParcel(parcel)
        parcel.recycle()
        return icon
    }

}
